File: include/apl/types/functions/builtins/mops/KeyBuiltin.java
package APL.types.functions.builtins.mops;

import APL.*;
import APL.errors.*;
import APL.types.*;
import APL.types.arrs.HArr;
import APL.types.functions.*;

import java.util.*;

public class KeyBuiltin extends Mop {
  @Override public std::string repr() {
    return "⌸";
  }

  public KeyBuiltin(Scope sc) {
    super(sc);
  }

  public Value call(Obj f, Value w, DerivedMop derv) {
    Obj o = callObj(f, w, derv);
    if (o instanceof Value) return (Value) o;
    throw new DomainError("Was expected to give array, got "+o.humanType(true), this);
  }
  public Obj callObj(Obj aa, Value w, DerivedMop derv) {
    if (aa instanceof APLMap) {
      if (w.rank > 1) {
        Value[] arr = new Value[w.ia];
        for (int i = 0; i < w.ia; i++) {
          arr[i] = (Value) ((APLMap) aa).getRaw(w.get(i));
        }
        return Arr.create(arr, w.shape);
      }
      return ((APLMap) aa).getRaw(w);
    }
    if (aa instanceof Fun) {
      int i = sc.IO;
      auto vals = new std::unordered_map<Value,std::vector<Value>>();
      auto order = new std::vector<Value>();
      for (Value v : w) {
        if (!vals.containsKey(v)) {
          auto l = new std::vector<Value>();
          l.add(Num.of(i));
          vals.put(v, l);
          order.add(v);
        } else {
          vals.get(v).add(Num.of(i));
        }
        i++;
      }
      auto res = new Value[order.size()];
      i = 0;
      for (Value c : order) {
        res[i++] = ((Fun)aa).call(c, Arr.create(vals.get(c)));
      }
      return new HArr(res);
    }
    throw new DomainError("⌸: ⍶ must be a function or a map, was "+aa.humanType(true), derv, aa);
  }

  public Value call(Obj aa, Value a, Value w, DerivedMop derv) {
    if (aa instanceof APLMap) {
      ((APLMap)aa).set(a, w);
      return w;
    }
    if (aa instanceof Fun) {
      Fun aaf = (Fun) aa;
      if (!Arrays.equals(a.shape, w.shape)) {
        if (a.rank != w.rank) throw new RankError("dyadic ⌸ expected ⍺ & ⍵ to have equal ranks ("+a.rank+" vs "+w.rank+")", derv, w);
        throw new LengthError("dyadic ⌸ expected ⍺ & ⍵ to have equal shapes ("+Main.formatAPL(a.shape)+" vs "+ Main.formatAPL(w.shape)+")", derv, w);
      }
      std::unordered_map<Value,std::vector<Value>> vals = new std::unordered_map<>();
     std::vector<Value> order = new std::vector<>();
      for (int i = 0; i < a.ia; i++) {
        Value k = w.get(i);
        Value v = a.get(i);
       std::vector<Value> curr = vals.get(k);
        if (curr == null) {
         std::vector<Value> newArr = new std::vector<>();
          vals.put(k, newArr);
          curr = newArr;
          order.add(k);
        }
        curr.add(v);
      }
      Value[] res = new Value[order.size()];
      for (int i = 0; i < order.size(); i++) {
        Value k = order.get(i);
        Value vs = Arr.create(vals.get(k));
        res[i] = aaf.call(k, vs);
      }
      return Arr.create(res);
    }
    throw new DomainError("⌸: ⍶ must be a function or a map, was "+aa.humanType(true), derv, aa);
  }
}

File: APL/types/functions/builtins/mops/KeyBuiltin.cpp
#include <APL/KeyBuiltin.h>
#include <APL/errors/DomainError.h>
#include <APL/types/APLMap.h>
#include <APL/types/Arr.h>
#include <APL/types/Fun.h>
#include <APL/types/Num.h>
#include <APL/errors/RankError.h>
#include <APL/Main.h>
#include <APL/errors/LengthError.h>

namespace APL::types::functions::builtins::mops
{
    // using namespace APL;
    // using namespace APL::errors;
    // using namespace APL::types;
    using HArr = APL::types::arrs::HArr;
    // using namespace APL::types::functions;

    std::string KeyBuiltin::repr()
    {
      return L"⌸";
    }

    KeyBuiltin::KeyBuiltin(std::shared_ptr<Scope> sc) : Mop(sc)
    {
    }

    std::shared_ptr<APL::types::Value> KeyBuiltin::call(std::shared_ptr<APL::types::Obj> f, std::shared_ptr<APL::types::Value> w, std::shared_ptr<DerivedMop> derv)
    {
      std::shared_ptr<APL::types::Obj> o = callObj(f, w, derv);
      if (std::dynamic_pointer_cast<APL::types::Value>(o) != nullptr)
      {
          return std::static_pointer_cast<APL::types::Value>(o);
      }
      throw DomainError(StringHelper::wstring_to_string(L"Was expected to give array, got " + o->humanType(true), shared_from_this()));
    }

    std::shared_ptr<APL::types::Obj> KeyBuiltin::callObj(std::shared_ptr<APL::types::Obj> aa, std::shared_ptr<APL::types::Value> w, std::shared_ptr<DerivedMop> derv)
    {
      if (std::dynamic_pointer_cast<APLMap>(aa) != nullptr)
      {
        if (w->rank > 1)
        {
          std::vector<std::shared_ptr<APL::types::Value>> arr(w->ia);
          for (int i = 0; i < w->ia; i++)
          {
            arr[i] = std::static_pointer_cast<APL::types::Value>((std::static_pointer_cast<APLMap>(aa))->getRaw(w[i]));
          }
          return Arr::create(arr, w->shape);
        }
        return (std::static_pointer_cast<APLMap>(aa))->getRaw(w);
      }
      if (std::dynamic_pointer_cast<Fun>(aa) != nullptr)
      {
        int i = sc->IO;
        auto vals = std::unordered_map<std::shared_ptr<APL::types::Value>, std::vector<std::shared_ptr<APL::types::Value>>>();
        auto order = std::vector<std::shared_ptr<APL::types::Value>>();
        for (auto v : w)
        {
          if (vals.find(v) == vals.end())
          {
            auto l = std::vector<std::shared_ptr<APL::types::Value>>();
            l.push_back(Num::of(i));
            vals.emplace(v, l);
            order.push_back(v);
          }
          else
          {
            vals[v].push_back(Num::of(i));
          }
          i++;
        }
        auto *res = std::vector<std::shared_ptr<APL::types::Value>>(order.size());
        i = 0;
        for (auto c : order)
        {
          res[i++] = (std::static_pointer_cast<Fun>(aa))->call(c, Arr::create(vals[c]));
        }
        return std::make_shared<HArr>(res);
      }
      throw DomainError(StringHelper::wstring_to_string(L"⌸: ⍶ must be a function or a map, was " + aa->humanType(true), derv, aa));
    }

    std::shared_ptr<APL::types::Value> KeyBuiltin::call(std::shared_ptr<APL::types::Obj> aa, std::shared_ptr<APL::types::Value> a, std::shared_ptr<APL::types::Value> w, std::shared_ptr<DerivedMop> derv)
    {
      if (std::dynamic_pointer_cast<APLMap>(aa) != nullptr)
      {
        (std::static_pointer_cast<APLMap>(aa))[a] = w;
        return w;
      }
      if (std::dynamic_pointer_cast<Fun>(aa) != nullptr)
      {
		std::shared_ptr<Fun> aaf = std::static_pointer_cast<Fun>(aa);
		if (!Arrays::equals(a->shape, w->shape))
		{
		  if (a->rank != w->rank)
		  {
			  throw RankError(StringHelper::wstring_to_string(L"dyadic ⌸ expected ⍺ & ⍵ to have equal ranks (" + std::to_wstring(a->rank) + L" vs " + std::to_wstring(w->rank) + L")", derv, w));
		  }
		  throw LengthError(StringHelper::wstring_to_string(L"dyadic ⌸ expected ⍺ & ⍵ to have equal shapes (" + Main::formatAPL(a->shape) + L" vs " + Main::formatAPL(w->shape) + L")", derv, w));
		}
		std::unordered_map<std::shared_ptr<APL::types::Value>, std::vector<std::shared_ptr<APL::types::Value>>> vals;
		std::vector<std::shared_ptr<APL::types::Value>> order;
		for (int i = 0; i < a->ia; i++)
		{
		  std::shared_ptr<APL::types::Value> k = w[i];
		  std::shared_ptr<APL::types::Value> v = a[i];
		  std::vector<std::shared_ptr<APL::types::Value>> curr = vals[k];
		  if (curr.empty())
		  {
			std::vector<std::shared_ptr<APL::types::Value>> newArr;
			vals.emplace(k, newArr);
			curr = newArr;
			order.push_back(k);
		  }
		  curr.push_back(v);
		}
		std::vector<std::shared_ptr<APL::types::Value>> res(order.size());
		for (int i = 0; i < order.size(); i++)
		{
		  std::shared_ptr<APL::types::Value> k = order[i];
		  std::shared_ptr<APL::types::Value> vs = Arr::create(vals[k]);
		  res[i] = aaf->call(k, vs);
		}
		return Arr::create(res);
	  }
	  throw DomainError(StringHelper::wstring_to_string(L"⌸: ⍶ must be a function or a map, was " + aa->humanType(true), derv, aa));
	}
}
